gh-132983: Add `compression.zstd` and Python tests

emmatyping opened the pull request:
The next step to merging the PEP 784 implementation, this PR introduces `compression.zstd` and tests for the code (which I have been using to test the C code so far).

Since test can be modified post-beta1, I'd like to focus on reviewing the API and Python implementation code. Obviously if there are important tests to add or change, that review would be valuable as well.





<!-- gh-issue-number: gh-132983 -->
* Issue: gh-132983
<!-- /gh-issue-number -->

emmatyping committed:
Add Python files

emmatyping committed:
Fix byteswarning in test

emmatyping committed:
Remove shape tests

emmatyping committed:
Make namedtuples dataclasses

AA-Turner commented on Lib/compression/zstd/__init__.py:
```diff
@@ -0,0 +1,293 @@
+    # From zstd.zstdfile
+    "open",
+    "ZstdFile",
+)
```
Consider sorting each section:

```suggestion
__all__ = (
    # compression.zstd
    "compressionLevel_values",
    "compress",
    "CParameter",
    "decompress",
    "DParameter",
    "finalize_dict",
    "get_frame_info",
    "Strategy",
    "train_dict",
    "zstd_support_multithread",

    # compression.zstd.zstdfile
    "open",
    "ZstdFile",

    # _zstd
    "get_frame_size",
    "zstd_version",
    "zstd_version_info",
    "ZstdCompressor",
    "ZstdDecompressor",
    "ZstdDict",
    "ZstdError",
)
```

Rogdham commented on Lib/compression/zstd/__init__.py:
```diff
@@ -0,0 +1,293 @@
+            if results:
+                break  # Leftover data is not a valid LZMA/XZ stream; ignore it.
+            else:
+                raise  # Error on the first iteration; bail out.
```
> Leftover data is not a valid LZMA/XZ stream; ignore it.

I think that it is not appropriate for Zstd (not that [I also think](https://github.com/python/cpython/issues/88300) that it is not appropriate for LZMA/XZ, so who knows).

Indeed, [the specification](https://github.com/facebook/zstd/blob/dev/doc/zstd_compression_format.md#frames) says:

> Zstandard compressed data is made of one or more frames

It does not say that random data can be added at the end.

---

Confirmation with `zstd` command:

```sh
$ (echo 'xxx' | zstd; echo 'yyy') > trailer_test.zstd
$ zstd -l trailer_test.zstd
Frames  Skips  Compressed  Uncompressed  Ratio  Check  Filename
Error: reached end of file with incomplete frame 
File "trailer_test.zstd" not compressed by zstd 
```
---

Issue with the current implementation:

```py
>>> trailer_test = compress(b"xxx") + b"yyy"
>>> decompress(trailer_test)  # should raise an exception
b'xxx'
```

emmatyping replied:
That's a really good point, and will likely require re-adding `EndlessZstdDecompressor` then (or just make `ZstdDecompressor` in `_zstd` equivalent to `EndlessZstdDecompressor` in pyzstd, which I think is probably the better solution).

Rogdham replied:
> will likely require re-adding EndlessZstdDecompressor

Maybe I'm missing something, but would the following solve the issue here?

```diff
diff --git a/Lib/compression/zstd/__init__.py b/Lib/compression/zstd/__init__.py
index affdf814f5..9d2a6bc08d 100644
--- a/Lib/compression/zstd/__init__.py
+++ b/Lib/compression/zstd/__init__.py
@@ -194,13 +194,7 @@ def decompress(data, zstd_dict=None, options=None):
     results = []
     while True:
         decomp = ZstdDecompressor(options=options, zstd_dict=zstd_dict)
-        try:
-            res = decomp.decompress(data)
-        except ZstdError:
-            if results:
-                break  # Leftover data is not a valid LZMA/XZ stream; ignore it.
-            else:
-                raise  # Error on the first iteration; bail out.
+        res = decomp.decompress(data)
         results.append(res)
         if not decomp.eof:
             raise ZstdError("Compressed data ended before the "
```

Your point is valid for my comment about `tarfile` below though (but it has less impact I would say).

> or just make ZstdDecompressor in _zstd equivalent to EndlessZstdDecompressor in pyzstd

My take on this: it is a good thing for advanced users to be able to decompress a single frame. Making that change would prevent it.

AA-Turner replied:
> My take on this: it is a good thing for advanced users to be able to decompress a single frame. Making that change would prevent it.

I don't know enough about the architecture of Zstandard to comment on the value of supporting single-frame operations. The design point I would make is that the stdlib will never cover every use-case, and shouldn't try to -- for advanced or more niche scenarios, PyPI packages will exist.

I would suggest ensuring what we have here covers the basics (& the PEP's specification) and isn't broken, and if needed defer the more complex features to the future. *'never is often better than \*right\* now'*!

A

Rogdham replied:
> The design point I would make is that the stdlib will never cover every use-case, and shouldn't try to -- for advanced or more niche scenarios, PyPI packages will exist.

You speak words of wisdom.

Still, if my patch above solves the issue at hand, I suggest to leave it as it is (which also avoids modifying the C implementation).

_For the record, we previously had the discussion about `ZstdDecompressor` vs `EndlessZstdDecompressor` at https://github.com/emmatyping/cpython/issues/7._

emmatyping replied:
I feel like supporting tarfile is important, but I agree that the fix suggested would resolve the issue here. For the tarfile issue, perhaps I can write a wrapper that handles multiple frames?

Then in the future we can decide to add `EndlessZstdDecompressor` which handles it natively in C code and use that.

Rogdham replied:
I like the idea of fixing the issue with `tarfile` in `r|zst` mode, without adding a new public api, which could be decided later.

Now that I'm thinking about this, we have the same issue with `ZstdDecompressor` usage in `zipfile`, don't we?

emmatyping replied:
> Now that I'm thinking about this, we have the same issue with ZstdDecompressor usage in zipfile, don't we?

Yes, we do. I believe that most tools write single frames but the appnote.txt does not specify so we should probably handle that as well.

emmatyping replied:
I think similarly though, zipfile support for multiple frames should go under "nice to have" but not block merging like mentioned in https://github.com/python/cpython/pull/133365#discussion_r2072698637

AA-Turner commented on Lib/compression/zstd/__init__.py:
```diff
@@ -0,0 +1,293 @@
+    min: int
+    max: int
+
+compressionLevel_values = _CompressionLevelValues(*_zstd._compressionLevel_values)
```
For Python code, we should adhere to PEP 8's guidance on naming where possible
```suggestion
compression_level_values = _CompressionLevelValues(*_zstd._compressionLevel_values)
```

emmatyping replied:
This one is a bit tricky, and I probably should have called it out. The `compressionLevel` name derives from the `ZSTD_c_compressionLevel` value in C. Since the min and max are already available through `CParameter.compressionLevel.bounds()`, perhaps it would be better to just set a value like `compressionLevel_default` or something and only include the result of `ZSTD_defaultCLevel()`? Or even set that on `CParameter.compressionLevel`?

AA-Turner replied:
> The `compressionLevel` name derives from the `ZSTD_c_compressionLevel` value in C.

As a general point, I still think we should avoid leaking details from the implementation language (C) into Python. Most/almost all of the users of `compression.zstd` won't have directly interacted with the C source of `compressionLevel`, and remembering why the naming is different will be an annoying quirk.

If you're strongly in favour of keeping the camelCase naming I won't block it, but I think we should think carefully here.

-----

> perhaps it would be better to just set a value like `compressionLevel_default`

`zstd.COMPRESS{,ION}_LEVEL_DEFAULT` could make sense, I guess the question is the value of exposing the number as a public attribute. A quick check shows that only `lzma` and `zlib` do so  (PRESET_DEFAULT & Z_DEFAULT_COMPRESSION), but `bz2` and `gzip` don't.

emmatyping replied:
> If you're strongly in favour of keeping the camelCase naming I won't block it, but I think we should think carefully here.

I'm not, I think it would be reasonable to rename to snake_case. @Rogdham interested in your thoughts though.

emmatyping replied:
Oh and
> zstd.COMPRESS{,ION}_LEVEL_DEFAULT could make sense, I guess the question is the value of exposing the number as a public attribute. A quick check shows that only lzma and zlib do so (PRESET_DEFAULT & Z_DEFAULT_COMPRESSION), but bz2 and gzip don't.

I think it is generally useful, people like to use these values when wrapping compression libraries in their own code I've seen across a few bugs. (it's nicer  to say `level=zstd.COMPRESSION_LEVEL_DEFAULT` than `level=None` in function definitions imo)

Rogdham replied:
> I think it would be reasonable to rename to snake_case. @Rogdham interested in your thoughts though

+1 on this (using snake case for constants is aligned with PEP8)

emmatyping replied:
Just pushed changes to reflect this :)

Rogdham commented on Lib/compression/zstd/__init__.py:
```diff
@@ -0,0 +1,293 @@
+    targetLength = _zstd._ZSTD_c_targetLength
+    strategy = _zstd._ZSTD_c_strategy
+
+    targetCBlockSize = _UnsupportedCParameter()
```
Is there any reason to keep this?

I introduced it in `pyzstd` to be able to have `targetCBlockSize` available or not depending on version of `zstd`, but here it is unconditional.

---

If you remove, remove `_UnsupportedCParameter` as well

emmatyping replied:
I guess if someone was to be hardcoding values I wanted to catch that it is unsupported at the CParameter level? But perhaps it is just better to leave that to the zstd error handling.

Rogdham replied:
Sorry I'm confused but I don't understand what you mean.

If you remove `targetCBlockSize` completely then users won't be able to use `CParameter.targetCBlockSize` as a key of the `options` dict (they would get `AttributeError`).

Or do you mean passing `130` as a key of the `options` dict? But I don't see how having `targetCBlockSize` or not changes anything.

emmatyping replied:
I meant the latter, and you're right that it doesn't add anything. I will remove this.

Rogdham commented on Lib/compression/zstd/__init__.py:
```diff
@@ -0,0 +1,293 @@
+    jobSize = _zstd._ZSTD_c_jobSize
+    overlapLog = _zstd._ZSTD_c_overlapLog
+
+    @functools.lru_cache(maxsize=None)
```
I had [an issue on `pyzstd`](https://github.com/Rogdham/pyzstd/pull/30) about using caching here.

Although I'm not sure that it is an issue, maybe it is a premature optimization and can be removed?

---

Same applies to `DParameter.bounds`.

AA-Turner replied:
I'd support removing the caching

emmatyping replied:
Yes I think for now I will remove it. Easier to add it than remove it.

AA-Turner commented on Lib/compression/zstd/__init__.py:
```diff
@@ -0,0 +1,293 @@
+class _CompressionLevelValues:
+    default: int
+    min: int
+    max: int
```
The only use of `dataclasses` is for two very simple immutable containers. For a slight boost to import performance, and to avoid implicitly exposing a ``dataclasses.is_dataclass()`` object, perhaps:

```suggestion
class _CLValues:
    __slots__ = 'default', 'min', 'max'

    def __init__(self, default, min, max):
        super().__setattr__('default', default)
        super().__setattr__('min', min)
        super().__setattr__('max', max)

    def __repr__(self):
        return (f'compression_level_values(default={self.default}, '
                f'min={self.min}, max={self.max})')

    def __setattr__(self, name, _):
        raise AttributeError(f"can't set attribute {name!r}")
```

AA-Turner commented on Lib/compression/zstd/__init__.py:
```diff
@@ -0,0 +1,293 @@
+class FrameInfo:
+    """A dataclass storing information about a Zstandard frame."""
+    decompressed_size: int
+    dictionary_id: int
```
Likewise:

```suggestion
class FrameInfo:
    """Information about a Zstandard frame."""
    __slots__ = 'decompressed_size', 'dictionary_id'

    def __init__(self, decompressed_size, dictionary_id):
        super().__setattr__('decompressed_size', decompressed_size)
        super().__setattr__('dictionary_id', dictionary_id)

    def __repr__(self):
        return (f'FrameInfo(decompressed_size={self.decompressed_size}, '
                f'dictionary_id={self.dictionary_id})')

    def __setattr__(self, name, _):
        raise AttributeError(f"can't set attribute {name!r}")
```

Rogdham commented on Lib/tarfile.py:
```diff
@@ -1817,11 +1829,13 @@ def open(cls, name=None, mode="r", fileobj=None, bufsize=RECORDSIZE, **kwargs):
            'w:gz'       open for writing with gzip compression
            'w:bz2'      open for writing with bzip2 compression
            'w:xz'       open for writing with lzma compression
+           'w:zst' open for writing with zstd compression
```
There seems to be indentation issues in the docstring (the `open` not aligned).

_Several occurrences_

AA-Turner commented on Lib/compression/zstd/__init__.py:
```diff
@@ -0,0 +1,293 @@
+    Parameter
+    frame_buffer: A bytes-like object. It should starts from the beginning of
+                  a frame, and needs to include at least the frame header (6 to
+                  18 bytes).
```
```suggestion
    *frame_buffer* is a bytes-like object. It should starts from the beginning of
    a frame, and needs to include at least the frame header (6 to 18 bytes).
```

AA-Turner commented on Lib/compression/zstd/__init__.py:
```diff
@@ -0,0 +1,293 @@
+    'decompressed_size' is the size in bytes of the data in the frame when
+    decompressed.
+
+    If decompressed_size is None, decompressed size is unknown.
```
```suggestion
    decompressed, or None when the decompressed size is unknown.
```

AA-Turner commented on Lib/compression/zstd/__init__.py:
```diff
@@ -0,0 +1,293 @@
+    """
+
+    ret_tuple = _zstd._get_frame_info(frame_buffer)
+    return FrameInfo(*ret_tuple)
```
```suggestion
    return FrameInfo(*_zstd._get_frame_info(frame_buffer))
```

AA-Turner commented on Lib/compression/zstd/__init__.py:
```diff
@@ -0,0 +1,293 @@
+    Parameters
+    samples:   An iterable of samples, a sample is a bytes-like object
+               represents a file.
+    dict_size: The dictionary's maximum size, in bytes.
```
```suggestion
    *samples* is an iterable of samples, where a sample is a bytes-like
    object representing a file.
    *dict_size* is the dictionary's maximum size, in bytes.
```

AA-Turner commented on Lib/compression/zstd/__init__.py:
```diff
@@ -0,0 +1,293 @@
+    """
+    # Check argument's type
+    if not isinstance(dict_size, int):
+        raise TypeError('dict_size argument should be an int object.')
```
Are int subclasses permitted?
```suggestion
    if not isinstance(dict_size, int):
        ds_cls = type(dict_size).__qualname__
        raise TypeError('dict_size must be an int object, not {ds_cls!r}.')
```

emmatyping replied:
I don't see why not. Anything that can be passed to `PyLong_AsSsize_t` should work.

AA-Turner commented on Lib/compression/zstd/__init__.py:
```diff
@@ -0,0 +1,293 @@
+    # samples_bytes: samples be stored concatenated in a single flat buffer.
+    # samples_size_list: a list of each sample's size.
+    # dict_size: size of the dictionary, in bytes.
+    dict_content = _train_dict(chunks, chunk_sizes, dict_size)
```
```suggestion
    samples = tuple(samples)
    chunks = b''.join(samples)
    chunk_sizes = tuple(map(_nbytes, samples))
    if not chunks:
        raise ValueError("samples contained no data; can't train dictionary.")
    dict_content = _train_dict(chunks, chunk_sizes, dict_size)
```

AA-Turner commented on Lib/compression/zstd/__init__.py:
```diff
@@ -0,0 +1,293 @@
+    dictionary can be a "raw content" dictionary, see is_raw parameter in
+    ZstdDict.__init__ method.
+
+    Parameters
```
Can you fix all the remaining 'Parameters' entries in docstrings? Check the result with pydoc/help.

emmatyping replied:
Hopefully my changes have fixed these properly.

AA-Turner commented on Lib/compression/zstd/__init__.py:
```diff
@@ -0,0 +1,293 @@
+
+    chunks = b''.join(chunks)
+    if not chunks:
+        raise ValueError("The samples are empty content, can't finalize dictionary.")
```
```suggestion
    samples = tuple(samples)
    chunks = b''.join(samples)
    chunk_sizes = tuple(map(_nbytes, samples))
    if not chunks:
        raise ValueError("samples contained no data; can't train dictionary.")
```

AA-Turner commented on Lib/compression/zstd/__init__.py:
```diff
@@ -0,0 +1,293 @@
+    # samples_bytes: samples be stored concatenated in a single flat buffer.
+    # samples_size_list: a list of each sample's size.
+    # dict_size: maximal size of the dictionary, in bytes.
+    # compression_level: compression level expected to use in production.
```
Lots of these comments don't add anything, can you remove all the redundant comments throughout?

Rogdham commented on Lib/tarfile.py:
```diff
@@ -399,7 +399,17 @@ def __init__(self, name, mode, comptype, fileobj, bufsize,
+                    raise CompressionError("compression.zstd module is not available") from None
+                if mode == "r":
+                    self.dbuf = b""
+                    self.cmp = zstd.ZstdDecompressor()
```
This will not work if the `.tar.zst` file is made of several frames.

Maybe this is not too bad  because `r:zst` will work (only `r|zst` will not).

---

Steps to reproduce below.

Create a `.tar.zst` file made of several frames:
```sh
$ head -c 200 /dev/urandom > file
$ tar cf file.tar file
$ (head -c 1000 file.tar|zstd; tail -c +1001 file.tar|zstd) >| file.tar.zst
```

Try decompressing:

```python
>>> t = tarfile.open("file.tar.zst", "r|zst")
>>> t.extractall('/path')
Traceback (most recent call last):
  File "<python-input-8>", line 1, in <module>
    t.extractall('/path')
    ~~~~~~~~~~~~^^^^^^^^^^
  File "/redacted/Lib/tarfile.py", line 2388, in extractall
    for member in members:
                  ^^^^^^^
  File "/redacted/Lib/tarfile.py", line 2882, in __iter__
    tarinfo = self.next()
  File "/redacted/Lib/tarfile.py", line 2733, in next
    raise ReadError("unexpected end of data")
tarfile.ReadError: unexpected end of data
```

gpshead replied:
is compressing a tar file as multiple independent concatenated frames a normal thing to do?  what do command line utilities do with that file?

Rogdham replied:
> is compressing a tar file as multiple independent concatenated frames a normal thing to do?

Usually, not. But nothing prevents people from doing it of course, and a typical usecase is random access support (e.g. assuming you store the positions of each file in the tar somewhere else, and do a frame every X files/length: when reading a single file in the tar you would be able to decompress just the right frame instead of the whole tar) although this requires advanced tooling.

> what do command line utilities do with that file?

They work transparently. For example, the `file.tar.zst` created in my last message would decompress with the `tar` command line just fine.

emmatyping replied:
I think perhaps we can put this under "nice to have" but not required for this PR? I am happy to make the changes in a follow up if there is time before b1.

Rogdham replied:
> perhaps we can put this under "nice to have" but not required for this PR

+1 on that, it works great when opening with `r:zst` (just `r|zst` is impacted), and the case should not happen often in practice

emmatyping replied:
I'll file a follow-up issue for this for tarfile and zipfile after this PR is merged.

AA-Turner commented on Lib/compression/zstd/__init__.py:
```diff
@@ -0,0 +1,293 @@
+                                  chunks, chunk_sizes,
+                                  dict_size, level)
+
+    return _zstd.ZstdDict(dict_content)
```
We should be consistent about using either ZstdDict or prefixing with `_zstd`
```suggestion
    return ZstdDict(dict_content)
```

Rogdham commented on Lib/test/test_zstd/test_core.py:
```diff
@@ -0,0 +1,2506 @@
+from test.support import _1M
+from test.support import Py_GIL_DISABLED
+
+zstd = import_module("compression.zstd")
```
Does not seem to be used?

emmatyping replied:
This is to skip this module when zstd is not available.

Rogdham commented on Lib/test/test_zstd/test_core.py:
```diff
@@ -0,0 +1,2506 @@
+from test.support import Py_GIL_DISABLED
+
+zstd = import_module("compression.zstd")
+_zstd = import_module("_zstd")
```
Is it good practice to test it there? Should the related tests be moved from `Lib` to `Modules`?

emmatyping replied:
Yeah the tests all live in `Lib/tests` except for `[_testcapi](https://github.com/python/cpython/tree/main/Modules/_testcapi), [_testinternalcapi](https://github.com/python/cpython/tree/main/Modules/_testinternalcapi), and [_testlimitedcapi](https://github.com/python/cpython/tree/main/Modules/_testlimitedcapi). At least as far as I am aware. I should probably move this import_module call first however.

AA-Turner commented on Lib/compression/zstd/__init__.py:
```diff
@@ -0,0 +1,293 @@
+    return _zstd.ZstdDict(dict_content)
+
+def compress(data, level=None, options=None, zstd_dict=None):
+    """Compress a block of data, return a bytes object of zstd compressed data.
```
Line length (the file would benefit from a once-over, several docstrings/lines are too long).

I think it's correct to say Zstandard compressed here, though I'm not sure how the project draws the distinction between the format/specification and the implementation.

```suggestion
    """Return Zstandard compressed *data* as bytes.
```

Rogdham commented on Lib/test/test_zstd/test_core.py:
```diff
@@ -0,0 +1,2506 @@
+TRAINED_DICT = None
+
+KB = 1024
+MB = 1024*1024
```
I see you tried to use `_KB` and `_MB` in tests already, maybe these line can be removed?

emmatyping replied:
These are actually unused! I went through and replaced all of the instances of 1024 with `_1K` while deleting these.

AA-Turner commented on Lib/compression/zstd/__init__.py:
```diff
@@ -0,0 +1,293 @@
+    """Compress a block of data, return a bytes object of zstd compressed data.
+
+    Refer to ZstdCompressor's docstring for a description of the
+    optional arguments *level*, *options*, and *zstd_dict*.
```
Avoid redirecting to another function, especially for a main entry point such as `compress`.

emmatyping replied:
Tried to make this a bit clearer, hopefully the new version is good!

AA-Turner commented on Lib/compression/zstd/__init__.py:
```diff
@@ -0,0 +1,293 @@
+    For incremental compression, use an ZstdCompressor instead.
+    """
+    comp = ZstdCompressor(level=level, options=options, zstd_dict=zstd_dict)
+    return comp.compress(data, ZstdCompressor.FLUSH_FRAME)
```
```suggestion
    return comp.compress(data, mode=ZstdCompressor.FLUSH_FRAME)
```

AA-Turner commented on Lib/compression/zstd/__init__.py:
```diff
@@ -0,0 +1,293 @@
+    """Decompress one or more frames of data.
+
+    Refer to ZstdDecompressor's docstring for a description of the
+    optional arguments *zstd_dict*, *options*.
```
Likewise, avoid redirecting to another function.

AA-Turner commented on Lib/compression/zstd/__init__.py:
```diff
@@ -0,0 +1,293 @@
+            res = decomp.decompress(data)
+        except ZstdError:
+            if results:
+                break  # Leftover data is not a valid LZMA/XZ stream; ignore it.
```
line length
```suggestion
                # Leftover data is not a valid LZMA/XZ stream; ignore it.
                break
```

AA-Turner commented on Lib/compression/zstd/__init__.py:
```diff
@@ -0,0 +1,293 @@
+        data = decomp.unused_data
+        if not data:
+            break
+    return b"".join(results)
```
Would it be more memory efficient to use a `bytearray`?

gpshead replied:
I doubt it.  Same mantra though: keep this simple as is and we can optimize later if found to be an issue.

Rogdham commented on Lib/test/test_zstd/test_core.py:
```diff
@@ -0,0 +1,2506 @@
+        with ZstdFile(C(), 'rb') as f:
+            with self.assertRaisesRegex(AttributeError, r'fileno'):
+                f.fileno()
+
```
I have recently added a [`test_name` in `pyzstd`]( https://github.com/Rogdham/pyzstd/blob/master/tests/test_zstd.py#L2707) if you want to include it here.

emmatyping replied:
Added with credit :)

AA-Turner commented on Lib/compression/zstd/__init__.py:
```diff
@@ -0,0 +1,293 @@
+            self.name,
+            zstd_version,
+        )
+        raise NotImplementedError(msg)
```
Is there a way to achieve this without a descriptor?

AA-Turner commented on Lib/compression/zstd/__init__.py:
```diff
@@ -0,0 +1,293 @@
+class CParameter(enum.IntEnum):
+    """Compression parameters"""
+
+    compressionLevel = _zstd._ZSTD_c_compressionLevel
```
How required is the camelCase naming? As per above, for new code we should avoid non-standard names.

AA-Turner commented on Lib/compression/zstd/__init__.py:
```diff
@@ -0,0 +1,293 @@
+    jobSize = _zstd._ZSTD_c_jobSize
+    overlapLog = _zstd._ZSTD_c_overlapLog
+
+    @functools.lru_cache(maxsize=None)
```
Why not use `functools.cache`?
```suggestion
    @functools.cache
```

AA-Turner commented on Lib/compression/zstd/__init__.py:
```diff
@@ -0,0 +1,293 @@
+
+    windowLogMax = _zstd._ZSTD_d_windowLogMax
+
+    @functools.lru_cache(maxsize=None)
```
Likewise
```suggestion
    @functools.cache
```

Is the `_get_param_bounds` function particularly expensive? This is the only use of `functools` in the module.

emmatyping replied:
I went ahead and deleted these, can always be added back if needed!

AA-Turner commented on Lib/compression/zstd/__init__.py:
```diff
@@ -0,0 +1,293 @@
+    def bounds(self):
+        """Return lower and upper bounds of a compression parameter, both inclusive."""
+        # 1 means compression parameter
+        return _zstd._get_param_bounds(1, self.value)
```
```suggestion
        return _zstd._get_param_bounds(is_compress=True, parameter=self.value)
```

Rogdham commented on Lib/compression/zstd/zstdfile.py:
```diff
@@ -0,0 +1,378 @@
+import builtins
```
I am not sure about conventions in CPython, but maybe rename this file to `_zstdfile.py` to convey the fact that `open` and `ZstdFile` should be imported from `compression.zstd` and not from `compression.zstd.zstdfile`?

Alternatively, move the contents into `__init__.py`. After all, this seems what is done e.g. in `lzma.py`.

emmatyping replied:
I'm hoping to keep the split of the implementation across files for maintainability purposes. I think it would be fine to rename this file to `_zstdfile.py`, but I don't have strong opinions of if it should be private.

AA-Turner replied:
We should have one canonical import location/path, so either `from compression.zstd.zstdfile import ZstdFile` or `from compression.zstd import ZstdFile`. I am +0 for the latter, and if so I'd suggest making `zstdfile` private as an implementation detail. It's always easier to make it public later than vice-versa.

emmatyping replied:
Yes agreed, I think I will make it private.

AA-Turner commented on Lib/compression/zstd/__init__.py:
```diff
@@ -0,0 +1,293 @@
+    def bounds(self):
+        """Return lower and upper bounds of a decompression parameter, both inclusive."""
+        # 0 means decompression parameter
+        return _zstd._get_param_bounds(0, self.value)
```
```suggestion
        return _zstd._get_param_bounds(is_compress=False, parameter=self.value)
```

AA-Turner commented on Lib/compression/zstd/__init__.py:
```diff
@@ -0,0 +1,293 @@
+    """Compression strategies, listed from fastest to strongest.
+
+    Note : new strategies _might_ be added in the future, only the order
+    (from fast to strong) is guaranteed.
```
This makes me feel a little uneasy, if we expose a numeric value (`Strategy` is an IntEnum) people will end up depending on it.
```suggestion
    Note that new strategies might be added in the future.
    Only the order (from fast to strong) is guaranteed,
    the numeric value might change.
```

emmatyping replied:
I don't love this either, but this comes from libzstd itself:

> /* note : new strategies _might_ be added in the future.
>                         Only the order (from fast to strong) is guaranteed */

https://facebook.github.io/zstd/doc/api_manual_v1.5.7.html

I'm not sure if there is a way we could expose this in a stable manner on top of an unstable interface from libzstd. I think the best we can do is document it loudly.

gpshead replied:
Even with plain numbers, the same has always been true in all compression libraries.  it is up to the underlying implementation as to what they mean, and some do change their meaning at times.  the ordering is indeed the only thing they all seem to retain.  ex: zlib and zlib-ng differ on some meanings as zlib-ng decided to try to match an un-named user intent on some numbers.  I like that zstd is at least giving names to some values.  that feels more explicit when read in code.  what happens under the hood is up to the library regardless.

AA-Turner commented on Lib/compression/zstd/__init__.py:
```diff
@@ -0,0 +1,293 @@
+# Set CParameter/DParameter types for validity check
+_zstd._set_parameter_types(CParameter, DParameter)
+
+zstd_support_multithread = CParameter.nbWorkers.bounds() != (0, 0)
```
```suggestion
zstd_support_multithreaded = CParameter.nbWorkers.bounds() != (0, 0)
```

gpshead replied:
is there a use for this attribute at all?  the norm for a CPython build is going to be that libzstd always supports multithreaded compression/decompression internally so I don't expect users will write code checking if it can or not?

emmatyping replied:
Yes I agree it is rather uncommon to have zstd without multithreading. I suppose it might be the case on emscripten? But that's niche enough I think deleting this entirely is reasonable.

emmatyping replied:
Deleted, moved the detection to tests where it is used in a few places. We should probably decide if we want to support building without multithreading period.

AA-Turner reviewed (changes requested):
Review of `Lib/compression/zstd/__init__.py`

Rogdham reviewed (commented):
This review is focused on changes from `pyzstd`.

My main issue is `decompress(compress(b"xxx") + b"yyy")` not raising an exception.

Rogdham commented on Lib/test/test_zstd/test_core.py:
```diff
@@ -0,0 +1,2506 @@
+        # larger than signed int, ValueError
+        d1 = d.copy()
+        d1[CParameter.ldmBucketSizeLog] = 2**31
+        self.assertRaises(ValueError, ZstdCompressor, d1)
```
```suggestion
        self.assertRaises(ValueError, ZstdCompressor, options=d1)
```

Rogdham commented on Lib/test/test_zstd/test_core.py:
```diff
@@ -0,0 +1,2506 @@
+        compress(b'', compressionLevel_values.min-1)
+
+        compress(b'', {CParameter.compressionLevel:compressionLevel_values.max+1})
+        compress(b'', {CParameter.compressionLevel:compressionLevel_values.min-1})
```
```suggestion
        compress(b'', options={CParameter.compressionLevel:compressionLevel_values.max+1})
        compress(b'', options={CParameter.compressionLevel:compressionLevel_values.min-1})
```

Rogdham commented on Lib/test/test_zstd/test_core.py:
```diff
@@ -0,0 +1,2506 @@
+            with self.assertRaises(ZstdError):
+                ZstdCompressor({CParameter.jobSize:4})
+            with self.assertRaises(ZstdError):
+                ZstdCompressor({CParameter.overlapLog:4})
```
```suggestion
            with self.assertRaises(ZstdError):
                ZstdCompressor(options={CParameter.nbWorkers:4})
            with self.assertRaises(ZstdError):
                ZstdCompressor(options={CParameter.jobSize:4})
            with self.assertRaises(ZstdError):
                ZstdCompressor(options={CParameter.overlapLog:4})
```

Rogdham commented on Lib/test/test_zstd/test_core.py:
```diff
@@ -0,0 +1,2506 @@
+                (r'Error when setting zstd compression parameter "windowLog", '
+                 r'it should \d+ <= value <= \d+, provided value is 100\. '
+                 r'\(zstd v\d\.\d\.\d, (?:32|64)-bit build\)')):
+            compress(b'', option)
```
```suggestion
            compress(b'', options=option)
```

Rogdham commented on Lib/test/test_zstd/test_core.py:
```diff
@@ -0,0 +1,2506 @@
+        pattern = r'Zstd compression parameter.*?"unknown parameter \(key %d\)"' \
+                  % KEY
+        with self.assertRaisesRegex(ZstdError, pattern):
+            ZstdCompressor(option)
```
```suggestion
            ZstdCompressor(options=option)
```

Rogdham commented on Lib/test/test_zstd/test_core.py:
```diff
@@ -0,0 +1,2506 @@
+    @classmethod
+    def setUpClass(cls):
+        options = {CParameter.checksumFlag:1}
+        c = ZstdCompressor(options)
```
```suggestion
        c = ZstdCompressor(options=options)
```

Rogdham commented on Lib/test/test_zstd/test_core.py:
```diff
@@ -0,0 +1,2506 @@
+            with ZstdFile(dst, "w", options={CParameter.checksumFlag:1}) as f:
+                f.write(raw_data)
+
+            comp = ZstdCompressor({CParameter.checksumFlag:1})
```
```suggestion
            comp = ZstdCompressor(options={CParameter.checksumFlag:1})
```

Rogdham reviewed (commented):
These extra comments are due to `options` being passed as positional arguments, but they are passed to `level` parameter instead of `options`.

I hope I did not miss any :sweat_smile:

emmatyping committed:
Apply suggestions from AA-Turner

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

emmatyping committed:
Clean up chunk calculations in train_dict

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

emmatyping committed:
Fix _CLValues instantiation

emmatyping committed:
More cleanup of train_/finalize_dict

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

emmatyping committed:
Have train_/finalize_dict take tuple not list

emmatyping committed:
Ensure trailing data raises errors

emmatyping committed:
Remove paramter bounds caching and unsupported...

parameters.

tomasr8 commented on Lib/compression/zstd/zstdfile.py:
```diff
@@ -0,0 +1,378 @@
+                        "options."
+                    )
+                )
+            if level:
```
```suggestion
            if level is not None:
```

tomasr8 commented on Lib/compression/zstd/zstdfile.py:
```diff
@@ -0,0 +1,378 @@
+            mode_code = _MODE_READ
+        elif mode in ("w", "wb", "a", "ab", "x", "xb"):
+            if not isinstance(level, (type(None), int)):
+                raise TypeError(("level argument should be an int object."))
```
```suggestion
                raise TypeError("level argument should be an int object.")
```

tomasr8 commented on Lib/compression/zstd/zstdfile.py:
```diff
@@ -0,0 +1,378 @@
+            if not isinstance(level, (type(None), int)):
+                raise TypeError(("level argument should be an int object."))
+            if not isinstance(options, (type(None), dict)):
+                raise TypeError(("options argument should be an dict object."))
```
```suggestion
                raise TypeError("options argument should be an dict object.")
```

tomasr8 commented on Lib/compression/zstd/zstdfile.py:
```diff
@@ -0,0 +1,378 @@
+            )
+            self._pos = 0
+        else:
+            raise ValueError("Invalid mode: {!r}".format(mode))
```
```suggestion
            raise ValueError(f"Invalid mode: {mode!r}")
```

tomasr8 commented on Lib/compression/zstd/zstdfile.py:
```diff
@@ -0,0 +1,378 @@
+        elif hasattr(filename, "read") or hasattr(filename, "write"):
+            self._fp = filename
+        else:
+            raise TypeError(("filename must be a str, bytes, file or PathLike object"))
```
```suggestion
            raise TypeError("filename must be a str, bytes, file or PathLike object")
```

emmatyping committed:
Use kwargs for code clarity

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

AA-Turner commented on Lib/compression/zstd/__init__.py:
```diff
@@ -116,7 +115,7 @@ def train_dict(samples, dict_size):
     samples = tuple(samples)
     chunks = b''.join(samples)
-    chunk_sizes = tuple(map(_nbytes, samples))
+    chunk_sizes = tuple(_nbytes(sample) for sample in samples)
```
? `map` is faster than a gencomp. The new listcomp inlining might be faster than `map`, though. (I used tuples to try and keep memory overhead low, but this might be a premature optimisation)

gpshead replied:
i'd stick with this as written, it is simpler.  leave optimizing for later if a profile shows anything as an issue.  (ex: the _zstd._train_dict code itself could even be changed to not take a chunks_sizes parameter)

emmatyping replied:
I generally see people know about comprehensions but don't always know about map, but map may be faster. The number of samples should be ~100-1000, with each maxing out at about 100KiB, so I think memory optimization is a good idea.

The vast majority of the time will be in actually training the dictionary in C code for cases where there are many samples, so I'm not sure an optimization for iteration on attribute access will be needed. I'm fine going back to map if that is preferred.

ref for FAQ for dictionaries discussing sample count and sizes (to be referenced in the docs probably): https://github.com/facebook/zstd/blob/e26dde3dcfc202b7e60403d90e5c13f0e0fbc7e9/lib/zdict.h#L117

tomasr8 commented on Lib/compression/zstd/zstdfile.py:
```diff
@@ -0,0 +1,378 @@
+
+    if "t" in mode:
+        if "b" in mode:
+            raise ValueError("Invalid mode: %r" % (mode,))
```
```suggestion
            raise ValueError(f"Invalid mode: {mode!r}")
```

emmatyping committed:
Clean up imports in zstd tests

tomasr8 commented on Lib/tarfile.py:
```diff
@@ -2006,12 +2024,48 @@ def xzopen(cls, name, mode="r", fileobj=None, preset=None, **kwargs):
 
+    @classmethod
+    def zstopen(cls, name, mode="r", fileobj=None, level=None, options=None,
+                 zstd_dict=None, **kwargs):
```
```suggestion
                zstd_dict=None, **kwargs):
```

tomasr8 commented on Lib/tarfile.py:
```diff
@@ -2006,12 +2024,48 @@ def xzopen(cls, name, mode="r", fileobj=None, preset=None, **kwargs):
+            if mode == 'r':
+                raise ReadError("not a zstd file") from e
+            raise
+        except:
```
We don't want to catch `BaseException`s here, right?
```suggestion
        except Exception:
```

emmatyping replied:
I think probably not, the code is based on what xzopen does above. But I think you are right that we should not catch `BaseException`.

emmatyping committed:
Use _1K instead of 1024 in tests

tomasr8 commented on Lib/tarfile.py:
```diff
@@ -2006,12 +2024,48 @@ def xzopen(cls, name, mode="r", fileobj=None, preset=None, **kwargs):
         "bz2": "bz2open",   # bzip2 compressed tar
-        "xz":  "xzopen"     # lzma compressed tar
+        "xz":  "xzopen",     # lzma compressed tar
+        "zst": "zstopen" # zstd compressed tar
```
```suggestion
        "xz":  "xzopen",    # lzma compressed tar
        "zst": "zstopen"    # zstd compressed tar
```

emmatyping committed:
Move compression.zstd.zstdfile to compression.zstd._zstdfile

This makes it private.

tomasr8 reviewed (commented):
Some nits, nothing major

gpshead commented on Lib/compression/zstd/__init__.py:
```diff
@@ -0,0 +1,258 @@
+"""Python bindings to Zstandard (zstd) compression library, the API style is
+similar to Python's bz2/lzma/zlib modules.
+"""
```
```suggestion
"""Python bindings to the Zstandard (zstd) compression library (RFC-8878)."""
```

emmatyping committed:
Change compressLevel_values to COMPRESSION_LEVEL_DEFAULT

emmatyping committed:
Fix tests for change in error message

tomasr8 commented on Lib/test/test_zstd/test_core.py:
```diff
@@ -0,0 +1,2475 @@
+        with self.assertRaisesRegex(ZstdError, pattern):
+            ZstdCompressor(option)
+
+    @unittest.skipIf(True,#not zstd_support_multithread,
```
Should this be uncommented?

emmatyping replied:
Yes, just fixed this!

tomasr8 commented on Lib/test/test_zstd/test_core.py:
```diff
@@ -0,0 +1,2475 @@
+        b = THIS_FILE_BYTES * (size // len(THIS_FILE_BYTES))
+
+        options = {CParameter.compressionLevel : 4,
+                  CParameter.nbWorkers : 2}
```
```suggestion
                   CParameter.nbWorkers : 2}
```

tomasr8 commented on Lib/test/test_zstd/test_core.py:
```diff
@@ -0,0 +1,2475 @@
+
+        # ZstdFile
+        with ZstdFile(io.BytesIO(), 'w',
+                      options=options) as f:
```
```suggestion
        with ZstdFile(io.BytesIO(), 'w', options=options) as f:
```

tomasr8 commented on Lib/test/test_zstd/test_core.py:
```diff
@@ -0,0 +1,2475 @@
+        dat1 += c.compress(THIS_FILE_BYTES[point:], c.FLUSH_BLOCK)
+        self.assertEqual(c.last_mode, c.FLUSH_BLOCK)
+        dat2 = c.flush()
+        pattern = r"Compressed data ended before the end-of-stream marker"
```
```suggestion
        pattern = "Compressed data ended before the end-of-stream marker"
```

gpshead commented on Lib/compression/zstd/__init__.py:
```diff
@@ -0,0 +1,244 @@
+    windowLogMax = _zstd._ZSTD_d_windowLogMax
+
+    def bounds(self):
+        """Return lower and upper bounds of a decompression parameter, both inclusive."""
```
indicate the return type, i'm presumably a tuple (lower, upper) of ints.

emmatyping replied:
Tried to make this a bit better, hopefully the new version looks ok?

emmatyping committed:
Make parameter names snake case

emmatyping committed:
Replace compressionLevel_values re-export with COMPRESSION_LEVEL_DEFAULT

emmatyping committed:
Move zstd_support_multithread to tests and rename

emmatyping committed:
Update module docstring for compression.zstd

Co-authored-by: Gregory P. Smith <greg@krypto.org>

emmatyping committed:
Clarify Strategy stability in docstring

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

emmatyping committed:
Fix formatting in tarfile

Co-authored-by: Tomas R. <tomas.roun8@gmail.com>

emmatyping committed:
Remove zstd_support_multithread from __all__

emmatyping committed:
Add test_name from upstream

Co-authored-by: Rogdham <contact@rogdham.net>

emmatyping committed:
Don't close tarfile if there is a BaseException

Co-authored-by: Tomas R. <tomas.roun8@gmail.com>

emmatyping committed:
Use options kwarg in tests

Co-authored-by: Rogdham <3994389+Rogdham@users.noreply.github.com>

emmatyping committed:
Use options kwarg in tests in more places

These were not shown because the flag names changed.

Co-authored-by: Rogdham <3994389+Rogdham@users.noreply.github.com>

emmatyping committed:
Adopt suggestions by Tomas R. for _zstdfile

Co-authored-by: Tomas R. <tomas.roun8@gmail.com>

emmatyping committed:
Formatting fixes in zstd tests

Co-authored-by: Tomas R. <tomas.roun8@gmail.com>

emmatyping committed:
Improve docstrings for (de)compress

emmatyping committed:
Fix some line length issues

AA-Turner commented on Lib/test/test_zstd/test_core.py:
There's only one test file, so there's no need to have a package here -- simpler to just have `test_zstd.py`

emmatyping replied:
Yeah, I think you're right. I was hoping to split it up so it wasn't so large, but didn't really have time to do that. I'll move it.

AA-Turner commented on Lib/compression/zstd/_zstdfile.py:
```diff
@@ -0,0 +1,378 @@
+import builtins
```
We're already importing `io`, so we can use `io.open` as ``builtins.open is io.open``.
```suggestion
```

AA-Turner commented on Lib/compression/zstd/_zstdfile.py:
```diff
@@ -0,0 +1,378 @@
+        if isinstance(filename, (str, bytes, PathLike)):
+            if "b" not in mode:
+                mode += "b"
+            self._fp = builtins.open(filename, mode)
```
We're already importing `io`, so we can use `io.open` as ``builtins.open is io.open``.
```suggestion
            self._fp = io.open(filename, f'{mode}b')
```

AA-Turner commented on Lib/compression/zstd/_zstdfile.py:
```diff
@@ -0,0 +1,378 @@
+    """
+
+    _READER_CLASS = _streams.DecompressReader
+
```
this is never changed/overriden
```suggestion
```

AA-Turner commented on Lib/compression/zstd/_zstdfile.py:
```diff
@@ -0,0 +1,378 @@
+        self._mode = mode_code
+
+        if self._mode == _MODE_READ:
+            raw = self._READER_CLASS(
```
per above
```suggestion
            raw = _streams.DecompressReader(
```

emmatyping committed:
Improve docstring on C/DParameter.bounds()

emmatyping committed:
Improve docstrings and formatting

emmatyping committed:
Add missing f string prefix

emmatyping committed:
Fix weird indent in _zstdfile.py

AA-Turner commented on Lib/compression/zstd/_zstdfile.py:
```diff
@@ -0,0 +1,378 @@
+            )
+            self._pos = 0
+        else:
+            raise ValueError(f"Invalid mode: {mode!r}")
```
```suggestion
        if options is not None and not isinstance(options, dict):
            raise TypeError("options must be a dict or None")
        mode = mode.removesuffix("b")  # handle rb, wb, xb, ab
        if mode == "r":
            if level is not None:
                raise TypeError("level is illegal in read mode")
            self._mode = _MODE_READ
        elif mode in {"w", "a", "x"}:
            if level is not None and not isinstance(level, int):
                raise TypeError("level must be int or None")
            self._mode = _MODE_WRITE
            self._compressor = ZstdCompressor(
                level=level, options=options, zstd_dict=zstd_dict
            )
            self._pos = 0
        else:
            raise ValueError(f"invalid mode: {mode!r}")
```

AA-Turner commented on Lib/compression/zstd/_zstdfile.py:
```diff
@@ -0,0 +1,378 @@
+        # File object
+        if isinstance(filename, (str, bytes, PathLike)):
+            if "b" not in mode:
+                mode += "b"
```
`b` can now always be appended
```suggestion
```

AA-Turner commented on Lib/compression/zstd/_zstdfile.py:
```diff
@@ -0,0 +1,378 @@
+            self._fp = builtins.open(filename, mode)
+            self._closefp = True
+        elif hasattr(filename, "read") or hasattr(filename, "write"):
+            self._fp = filename
```
This should also check the mode?
```suggestion
        elif ((mode == 'r' and hasattr(filename, "read"))
              or (mode != 'r' and hasattr(filename, "write"))):
            self._fp = filename
```

AA-Turner commented on Lib/compression/zstd/_zstdfile.py:
```diff
@@ -0,0 +1,378 @@
+        elif hasattr(filename, "read") or hasattr(filename, "write"):
+            self._fp = filename
+        else:
+            raise TypeError("filename must be a str, bytes, file or PathLike object")
```
```suggestion
            raise TypeError("filename must be a file-like object "
                            "or a str, bytes, or PathLike object")
```

AA-Turner commented on Lib/compression/zstd/_zstdfile.py:
```diff
@@ -0,0 +1,378 @@
+
+    def __init__(
+        self,
+        filename,
```
Given it can accept file pointers, `filename` isn't quite accurate. Also make it positional-only
```suggestion
        file, /,
```

AA-Turner commented on Lib/compression/zstd/_zstdfile.py:
```diff
@@ -0,0 +1,378 @@
+        May be called more than once without error. Once the file is
+        closed, any other operation on it will raise a ValueError.
+        """
+        # Nop if already closed
```
prefer either capitalised NOP or fully spelt out, but in this case I think the comment can be removed
```suggestion
```

AA-Turner commented on Lib/compression/zstd/_zstdfile.py:
```diff
@@ -0,0 +1,378 @@
+        """Flush and close the file.
+
+        May be called more than once without error. Once the file is
+        closed, any other operation on it will raise a ValueError.
```
```suggestion
        May be called multiple times. Once the file has been closed,
        any other operation on it will raise ValueError.
```

AA-Turner commented on Lib/compression/zstd/_zstdfile.py:
```diff
@@ -0,0 +1,378 @@
+            return
+        try:
+            if self._mode == _MODE_READ:
+                if hasattr(self, "_buffer") and self._buffer:
```
we always have `self._buffer` if mode is MODE_READ, the guard isn't needed.
```suggestion
                if self._buffer:
```

If you want to keep the guard, simplify it to `getattr(self, '_buffer', None)`, but also ensure `self._compressor` below is guarded.

emmatyping replied:
When I remove the guard, `FileTestCase.test_init_bad_check` fails:

```
test_init_bad_check (test.test_zstd.test_core.FileTestCase.test_init_bad_check) ... Warning -- Unraisable exception
Exception ignored while finalizing file <compression.zstd._zstdfile.ZstdFile object at 0x7fcac7a7cc30>:
Traceback (most recent call last):
  File "/home/emma/cpython/Lib/compression/zstd/_zstdfile.py", line 117, in close
    self._buffer.close()
AttributeError: 'ZstdFile' object has no attribute '_buffer'
```

The getattr replacement is smart though, I will definitely use that.

emmatyping replied:
Right I just thought through when that will be the case: if `__init__` fails before `self._buffer` is initialized, then it will have this issue. I'll add a defensive `self._buffer = None` to the beginning to ensure that case won't happen (I think I am going to leave the check in `close()` to be defensive though).

AA-Turner commented on Lib/compression/zstd/_zstdfile.py:
```diff
@@ -0,0 +1,378 @@
+            decompression.
+        """
+        self._fp = None
+        self._closefp = False
```
throughout: rename to `_close_fp`?

AA-Turner commented on Lib/compression/zstd/_zstdfile.py:
```diff
@@ -0,0 +1,378 @@
+        else:
+            # accept any data that supports the buffer protocol
+            data = memoryview(data)
+            length = data.nbytes
```
can this use the nbytes helper?

emmatyping replied:
Unfortunately that would lead to a circular import. I could split `_nbytes` out into it's own file I guess?

Rogdham replied:
Would moving the helper into `_zstdfile` solve the circular import?

AA-Turner replied:
Moving to this file sounds good

AA-Turner commented on Lib/compression/zstd/_zstdfile.py:
```diff
@@ -0,0 +1,378 @@
+        if self._mode == _MODE_READ:
+            return
+        self._check_not_closed()
+        if mode not in (self.FLUSH_BLOCK, self.FLUSH_FRAME):
```
```suggestion
        if mode not in {self.FLUSH_BLOCK, self.FLUSH_FRAME}:
```

AA-Turner commented on Lib/compression/zstd/_zstdfile.py:
```diff
@@ -0,0 +1,378 @@
+    def flush(self, mode=FLUSH_BLOCK):
+        """Flush remaining data to the underlying stream.
+
+        The mode argument can be ZstdFile.FLUSH_BLOCK, ZstdFile.FLUSH_FRAME.
```
```suggestion
        The mode argument can be FLUSH_BLOCK or FLUSH_FRAME.
```

AA-Turner commented on Lib/compression/zstd/_zstdfile.py:
```diff
@@ -0,0 +1,378 @@
+        if mode not in (self.FLUSH_BLOCK, self.FLUSH_FRAME):
+            raise ValueError("mode argument wrong value, it should be "
+                             "ZstdCompressor.FLUSH_FRAME or "
+                             "ZstdCompressor.FLUSH_BLOCK.")
```
```suggestion
            raise ValueError("invalid mode argument, expected either "
                             "ZstdCompressor.FLUSH_FRAME or "
                             "ZstdCompressor.FLUSH_BLOCK")
```

AA-Turner commented on Lib/compression/zstd/_zstdfile.py:
```diff
@@ -0,0 +1,378 @@
+        This method does nothing in reading mode.
+        """
+        if self._mode == _MODE_READ:
+            return
```
What do the other compression modules do on `flush()` to a file in reading mode? Wondering if it is better to raise an exception here.

emmatyping replied:
Most of them don't implement it! gzip does nothing in read mode: https://github.com/python/cpython/blob/2bbcaedb75942389dacb51866948f40de5951c9c/Lib/gzip.py#L406-L412

AA-Turner replied:
Ok

AA-Turner commented on Lib/compression/zstd/_zstdfile.py:
```diff
@@ -0,0 +1,378 @@
+    def __next__(self):
+        ret = self._buffer.readline()
+        if ret:
+            return ret
```
```suggestion
        if ret := self._buffer.readline():
            return ret
```

AA-Turner reviewed (commented):
Comments on Lib/compression/zstd/_zstdfile.py

AA-Turner commented on Lib/compression/zstd/__init__.py:
```diff
@@ -0,0 +1,259 @@
+
+
+class CParameter(enum.IntEnum):
+    """Compression parameters"""
```
bikeshed: perhaps `CompressionOptions` and `DecompressionOptions`? These are only used as keys to the `options` dict, as far as I can tell.

It would be good to expand `C` to `Compress(ion)?` though (regardless of Parameter/Options/etc) if we  it makes it more obvious what this is for. When first skimming this I assumed it had to do with the C language, rather than C for Compression. Same applies for `D`.

emmatyping replied:
I guess my main concern is that `CompressionOptions.compression_level` seems rather long to type. One thing I discussed with @Rogdham was setting CParameter/DParameter members as module attributes. Past discussion here: https://github.com/emmatyping/cpython/issues/5

Rogdham replied:
> It would be good to expand C to Compress(ion)?

I am sightly in favor of expending the name, to avoid any confusion.

> Explicit is better than implicit.
> ‒ The Zen of Python, by Tim Peters

People finding it too long in their code can always do
```python
from compression.zstd import CompressionParameter as CParam
```

---

> Parameter/Options/etc

I would be in favor of “Parameter” here.

In the [zstd manual](https://facebook.github.io/zstd/zstd_manual.html), although they are naming the thing `ZSTD_cParameter`, they are speaking about “compression parameters” when using plain English.

Also personally, I think of “options” as the whole dict whereas “parameter” is a key of the dict. I am probably biased on this however.

AA-Turner replied:
`CompressionParameter` and `DecompressionParameter` are fine with me. If I understand, these are advanced options anyway, so making them a little more verbose isn't the worst thing.

emmatyping replied:
Yeah I agree, having them be verbose is fine since they are more advanced.

emmatyping replied:
Renamed to `CompressionParameter` and `DecompressionParameter`.

emmatyping committed:
Use io.open instead of builtins.open

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

emmatyping committed:
Remove _READER_CLASS from ZstdFile

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

emmatyping committed:
Adopt many suggestions from AA-Turner for ZstdFile

* rename filename argument to file
* improve __init__ mode and argument checking
* docstring and error rewording
* renamed self._closefp to self._close_fp
* removed mode_code from __init__
* removed unneeded self._READER_CLASS

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

emmatyping committed:
Set self._buffer to None

emmatyping committed:
Move _nbytes to _zstdfile.py

emmatyping committed:
Move test_zstd to file

emmatyping committed:
Rename C/DParameter to (De)CompressionParameter

AA-Turner committed:
regen clinic

emmatyping committed:
Fix whitespace issue

AA-Turner commented on Lib/compression/zstd/__init__.py:
```diff
@@ -0,0 +1,252 @@
+
+        Both lower and upper bounds are inclusive.
+        """
+        return _zstd._get_param_bounds(is_compress=True, parameter=self.value)
```
Minor, but we only need `is_compress` to be kwarg here. What do you think about c8db13de92a3f534a6c6f8777440b0d98b9d788a?

emmatyping replied:
I think it is a good idea. Feel free to push that to the branch.

emmatyping replied:
Just pushed this up myself.

emmatyping committed:
Remove makefile test dir

AA-Turner committed:
swap order of parameters in _get_param_bounds

AA-Turner committed:
Sort imports

AA-Turner committed:
Improve docstrings

AA-Turner committed:
Remove comments

AA-Turner committed:
Remove unused private variables

emmatyping committed:
Merge branch 'main' into 3.14-zstd-python-code

AA-Turner committed:
Misc changes (positional-only, style, error messages)

AA-Turner committed:
whitespace

AA-Turner committed:
Remove _set_parameter_types

Python, in general, works on protocols rather than nominal or specific types. I think this check is overly restrictive, with only minor benefit. Users wanting validation can use static type checkers.

emmatyping committed:
Revert "Remove _set_parameter_types"

This reverts commit bf4b07daade1a55da151db571fc581d80899e8db.

Checking the type of parameters is important to avoid confusing error
messages.

AA-Turner reviewed (approved):
This is mergeable, we can tweak the few remaining items (eg. tarfile, parameter type checks, perhaps tests) after 3.14 beta 1.

Thanks everyone, and especially thank you @Rogdham for the valuable feedback & context from `pyzstd`.

A

merwok commented on Lib/tarfile.py:
```diff
@@ -2006,12 +2024,48 @@ def xzopen(cls, name, mode="r", fileobj=None, preset=None, **kwargs):
         "bz2": "bz2open",   # bzip2 compressed tar
-        "xz":  "xzopen"     # lzma compressed tar
+        "xz":  "xzopen",    # lzma compressed tar
+        "zst": "zstopen"    # zstd compressed tar
```
Trailing comma would have been nice here 🙂 
(noting this in case someone does another cleanup PR, but not to request one just for this)

emmatyping replied:
will add this to https://github.com/python/cpython/pull/133547, thanks!
